src/hera/workflows/_runner/script_annotations_util.py

```diff
@@ -95,40 +95,54 @@ def get_annotated_output_param(param_annotation: Parameter) -> Path:
     return path
 
 
-def get_annotated_artifact_value(artifact_annotation: Artifact) -> Union[Path, Any]:
+def get_annotated_output_artifact(artifact_annotation: Artifact) -> Path:
+    if artifact_annotation.path:
+        path = Path(artifact_annotation.path)
+    else:
+        path = _get_outputs_path(artifact_annotation)
+    # Automatically create the parent directory (if required)
+    path.parent.mkdir(parents=True, exist_ok=True)
+    return path
+
+
+def get_annotated_artifact_value(artifact_annotation: Artifact, func_param_annotation: type) -> Union[Path, Any]:
     """Get the value of the given Artifact annotation.
 
     If the artifact is an output, return the path it will write to.
-    If the artifact is an input, return the loaded value (json, path or string) using its ArtifactLoader.
+    If the artifact is an input, return the  loaded value (json obj, path or string)
+    using its ArtifactLoader, unless ArtifactLoader is a Callable, in which case it
+    is used to load the value.
 
     As Artifacts are always Annotated in function parameters, we don't need to consider
     the `kwargs` or the function parameter name.
     """
     if artifact_annotation.output:
-        if artifact_annotation.path:
-            path = Path(artifact_annotation.path)
-        else:
-            path = _get_outputs_path(artifact_annotation)
-        # Automatically create the parent directory (if required)
-        path.parent.mkdir(parents=True, exist_ok=True)
-        return path
+        return get_annotated_output_artifact(artifact_annotation)
 
     if not artifact_annotation.path:
         # Path is added to the spec automatically when built. As it isn't present in the annotation itself,
         # we need to add it back in for the runner.
         artifact_annotation.path = artifact_annotation._get_default_inputs_path()
 
-    if artifact_annotation.loader == ArtifactLoader.json.value:
-        path = Path(artifact_annotation.path)
-        with path.open() as f:
-            return json.load(f)
+    actual_type = unwrap_annotation(func_param_annotation)
 
-    if artifact_annotation.loader == ArtifactLoader.file.value:
-        path = Path(artifact_annotation.path)
+    if artifact_annotation.loader is None:
+        if issubclass(actual_type, Path):
+            return artifact_annotation.path
+
+    path = Path(artifact_annotation.path)
+    if artifact_annotation.loader == ArtifactLoader.file:
         return path.read_text()
 
-    if artifact_annotation.loader is None:
-        return artifact_annotation.path
+    if artifact_annotation.loader == ArtifactLoader.json or artifact_annotation.loader is None:
+        with path.open() as f:
+            return json.load(f)
```

Comment:
This is a change of behaviour for when loader is `None`, correct? We never previously decoded it. Won't that risk breaking people?

Reply:
This would only be breaking if someone is using a `str` type input with a loader of `None`, so they want the Path but as a `str`, which I would consider a rare use case. I was considering if/how to use experimental flags for this feature, which I think make sense as a way to deprecate the old `ArtifactLoader` enum, or at least its `None` behaviour, so we make it require a `Path`, or add a new `ArtifactLoader.Path`. The main loader/dumper feature would not need an experimental flag.